pkg/ephemeral/spdz.go

```diff
@@ -247,3 +330,55 @@ func (s *SPDZEngine) writeIPFile(path string, addr string, parties int32) error
 	s.logger.Infow("Writing to IPFile: ", "path", path, "content", addrs, "proxy address", addr, "parties", parties)
 	return ioutil.WriteFile(path, data, 0644)
 }
+
+// preparePlayerData Returns the directories for the supported protocol's preprocessing data. It therefore creates
+// the required directories and writes the mac keys to the files expected by SPDZ.
+func preparePlayerData(conf *SPDZEngineTypedConfig) (map[castor.SPDZProtocol]string, error) {
+	playerDataDirs := make(map[castor.SPDZProtocol]string)
+	for _, p := range castor.SupportedSpdzProtocols {
+		path, err := createPlayerDataForProtocol(p, conf)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create preprocessing data directories: %v", err)
+		}
+		playerDataDirs[p] = path
+
+	}
+	return playerDataDirs, nil
+}
+
+func createPlayerDataForProtocol(p castor.SPDZProtocol, conf *SPDZEngineTypedConfig) (string, error) {
+	var playerDataDir, macKey string
+	switch p {
+	case castor.SpdzGfp:
+		playerDataDir = fmt.Sprintf("%s/%d-%s-%d/",
+			conf.PrepFolder, conf.PlayerCount, castor.SpdzGfp.Shorthand, conf.Prime.BitLen())
+		macKey = conf.GfpMacKey.String()
+	case castor.SpdzGf2n:
+		playerDataDir = fmt.Sprintf("%s/%d-%s-%d/",
+			conf.PrepFolder, conf.PlayerCount, castor.SpdzGf2n.Shorthand, conf.Gf2nBitLength)
+		macKey = conf.Gf2nMacKey
+	default:
+		panic("Unsupported SpdzProtocol " + p.Descriptor)
+	}
+	err := os.MkdirAll(playerDataDir, 0755)
+	if err != nil && !os.IsNotExist(err) {
+		return "", fmt.Errorf("error creating directory path: %v", err)
+	}
+	macKeyFileName := fmt.Sprintf("Player-MAC-Keys-%s-P%d", p.Shorthand, conf.PlayerID)
+	err = writeMacKey(playerDataDir+macKeyFileName, conf.PlayerCount, macKey)
+	if err != nil && !os.IsNotExist(err) {
```

Comment:
Why is `IsNotExist` fine here?

Reply:
It is not - most likely copy paste error. 
But this pointed me to an error in `writeMacKey(string, int32, string)` where the mac key was was not properly overwritten in case the file exists and previous data is longer than the new content.